From 1d0a0f87d070bb46542dd313c753805f035ef79a Mon Sep 17 00:00:00 2001
From: q66 <q66@chimera-linux.org>
Date: Sun, 23 Mar 2025 15:01:57 +0100
Subject: [PATCH] use openssl for hashing

---
 Makefile |  6 +++---
 main.c   | 48 ++++++++++++++++++++++++++++++++----------------
 2 files changed, 35 insertions(+), 19 deletions(-)

diff --git a/Makefile b/Makefile
index fb3d8d0..5e102ed 100644
--- a/Makefile
+++ b/Makefile
@@ -7,8 +7,8 @@ LDFLAGS =
 devmapper-CFLAGS = $(shell $(PKG_CONFIG) --cflags devmapper)
 devmapper-LDFLAGS = $(shell $(PKG_CONFIG) --libs devmapper)
 
-libmd-CFLAGS = $(shell $(PKG_CONFIG) --cflags libmd)
-libmd-LDFLAGS = $(shell $(PKG_CONFIG) --libs libmd)
+libcrypto-CFLAGS = $(shell $(PKG_CONFIG) --cflags libcrypto)
+libcrypto-LDFLAGS = $(shell $(PKG_CONFIG) --libs libcrypto)
 
 blkid-CFLAGS = $(shell $(PKG_CONFIG) --cflags blkid)
 blkid-LDFLAGS = $(shell $(PKG_CONFIG) --libs blkid)
@@ -27,4 +27,4 @@ clean:
 	rm -f make-dynpart-mappings
 
 make-dynpart-mappings: main.c 3rdparty/metadata_format.h
-	$(CC) -o make-dynpart-mappings main.c $(CFLAGS) $(LDFLAGS) $(CONFIG_DEFS) $(devmapper-CFLAGS) $(devmapper-LDFLAGS) $(libmd-CFLAGS) $(libmd-LDFLAGS) $(blkid-CFLAGS) $(blkid-LDFLAGS) -I3rdparty
+	$(CC) -o make-dynpart-mappings main.c $(CFLAGS) $(LDFLAGS) $(CONFIG_DEFS) $(devmapper-CFLAGS) $(devmapper-LDFLAGS) $(libcrypto-CFLAGS) $(libcrypto-LDFLAGS) $(blkid-CFLAGS) $(blkid-LDFLAGS) -I3rdparty
diff --git a/main.c b/main.c
index c938945..4a10e55 100644
--- a/main.c
+++ b/main.c
@@ -13,11 +13,14 @@
 
 #include <blkid/blkid.h>
 #include <libdevmapper.h>
-#include <sha2.h>
+#include <openssl/evp.h>
 
 #include "metadata_format.h"
 
 #define LINEAR_PARAMS_SIZE 256
+#define SHA256_DIGEST_LENGTH 32
+
+static EVP_MD_CTX *sha2;
 
 static const uint8_t zero[32] = {
 	0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
@@ -33,8 +36,8 @@ static const uint8_t zero[32] = {
  */
 
 LpMetadataGeometry *validateMetadataGeometry(LpMetadataGeometry *geom, const char *scope) {
-	SHA2_CTX sha2;
 	uint8_t checksum[SHA256_DIGEST_LENGTH];
+	unsigned int mdlen = 0;
 
 	if (le32toh(geom->magic) != LP_METADATA_GEOMETRY_MAGIC) {
 		fprintf(stderr, "%s: Invalid metadata geometry magic number (expected 0x%X, got 0x%X)\n", scope, LP_METADATA_GEOMETRY_MAGIC, le32toh(geom->magic));
@@ -47,11 +50,13 @@ LpMetadataGeometry *validateMetadataGeometry(LpMetadataGeometry *geom, const cha
 	}
 
 	// Calculate the checksum if the checksum field was zero
-	SHA256Init(&sha2);
-	SHA256Update(&sha2, (const uint8_t *) geom, offsetof(LpMetadataGeometry, checksum));
-	SHA256Update(&sha2, zero, 32);
-	SHA256Update(&sha2, &geom->checksum[32], geom->struct_size - offsetof(LpMetadataGeometry, checksum) - 32);
-	SHA256Final(checksum, &sha2);
+	EVP_DigestUpdate(sha2, (const uint8_t *) geom, offsetof(LpMetadataGeometry, checksum));
+	EVP_DigestUpdate(sha2, zero, 32);
+	EVP_DigestUpdate(sha2, &geom->checksum[32], geom->struct_size - offsetof(LpMetadataGeometry, checksum) - 32);
+	if (!EVP_DigestFinal(sha2, checksum, &mdlen)) {
+		fprintf(stderr, "%s: could not finalize sha256 digest\n", scope);
+		return NULL;
+	}
 
 	if (memcmp(checksum, geom->checksum, SHA256_DIGEST_LENGTH)) {
 		fprintf(stderr, "%s: Metadata geometry integrity check failed\n", scope);
@@ -72,8 +77,8 @@ LpMetadataGeometry *validateMetadataGeometry(LpMetadataGeometry *geom, const cha
 }
 
 LpMetadataHeader *validateMetadataHeader(LpMetadataHeader *header, const char *scope) {
-	SHA2_CTX sha2;
 	uint8_t checksum[SHA256_DIGEST_LENGTH];
+	unsigned int mdlen = 0;
 
 	if (le32toh(header->magic) != LP_METADATA_HEADER_MAGIC) {
 		fprintf(stderr, "%s: Invalid magic number for the metadata header (expected 0x%X, got 0x%X)\n", scope, LP_METADATA_HEADER_MAGIC, le32toh(header->magic));
@@ -100,11 +105,13 @@ LpMetadataHeader *validateMetadataHeader(LpMetadataHeader *header, const char *s
 	}
 
 	// Calculate the checksum if the checksum field was zero
-	SHA256Init(&sha2);
-	SHA256Update(&sha2, (const uint8_t *) header, offsetof(LpMetadataHeader, header_checksum));
-	SHA256Update(&sha2, zero, 32);
-	SHA256Update(&sha2, &header->header_checksum[32], header->header_size - offsetof(LpMetadataHeader, header_checksum) - 32);
-	SHA256Final(checksum, &sha2);
+	EVP_DigestUpdate(sha2, (const uint8_t *) header, offsetof(LpMetadataHeader, header_checksum));
+	EVP_DigestUpdate(sha2, zero, 32);
+	EVP_DigestUpdate(sha2, &header->header_checksum[32], header->header_size - offsetof(LpMetadataHeader, header_checksum) - 32);
+	if (!EVP_DigestFinal(sha2, checksum, &mdlen)) {
+		fprintf(stderr, "%s: could not finalize sha256 digest\n", scope);
+		return NULL;
+	}
 
 	if (memcmp(checksum, header->header_checksum, SHA256_DIGEST_LENGTH)) {
 		fprintf(stderr, "%s: Metadata header integrity check failed\n", scope);
@@ -116,9 +123,11 @@ LpMetadataHeader *validateMetadataHeader(LpMetadataHeader *header, const char *s
 		return NULL;
 	}
 
-	SHA256Init(&sha2);
-	SHA256Update(&sha2, (const void *) &((const uint8_t *) header)[le32toh(header->header_size)], le32toh(header->tables_size));
-	SHA256Final(checksum, &sha2);
+	EVP_DigestUpdate(sha2, (const void *) &((const uint8_t *) header)[le32toh(header->header_size)], le32toh(header->tables_size));
+	if (!EVP_DigestFinal(sha2, checksum, &mdlen)) {
+		fprintf(stderr, "%s: could not finalize sha256 digest\n", scope);
+		return NULL;
+	}
 
 	if (memcmp(checksum, header->tables_checksum, SHA256_DIGEST_LENGTH)) {
 		fprintf(stderr, "%s: Metadata tables integrity check failed\n", scope);
@@ -262,6 +271,12 @@ int main(int argc, char **argv) {
 	if (argc >= 3)
 		slotN = strtol(argv[2], NULL, 0);
 
+	sha2 = EVP_MD_CTX_new();
+	if (!sha2 || !EVP_DigestInit_ex(sha2, EVP_sha256(), NULL)) {
+		fprintf(stderr, "Could not init sha256 digest\n");
+		return 1;
+	}
+
 	geom = malloc(LP_METADATA_GEOMETRY_SIZE * 2);
 
 	// Seek to the beginning of the metadata
@@ -334,6 +349,7 @@ int main(int argc, char **argv) {
 	free(header);
 	free(geom);
 	free(superPartition);
+	EVP_MD_CTX_free(sha2);
 
 	return 0;
 }
-- 
2.48.1
